{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6619239-55d4-4004-af66-e5344f8fb44c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!git --version\n",
    "#!git clone https://github.com/evdokiazorina/homeworks.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee57826-aeed-42a6-b168-0c5f5fe1c1e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b0fb34-74e8-4dcd-8baa-d7e461306de8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 1. Получить список функций белка P04637 (p53)\n",
    "url = \"https://www.ebi.ac.uk/proteins/api/proteins/P04637\"\n",
    "response = requests.get(url, headers={\"Accept\": \"application/json\"})\n",
    "\n",
    "data = response.json()\n",
    "for feature in data[\"features\"]:\n",
    "    feature_type = feature.get(\"type\", \"\")\n",
    "    description = feature.get(\"description\", \"\")\n",
    "    print(f\"{feature_type}: {description}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76415b98-39cb-41eb-93de-e4e23a56430d",
   "metadata": {},
   "outputs": [],
   "source": [
    "###2. Получите последовательности белков P05067 (APP) и P10636 (Tau). Сравните их длины. Какой белок длиннее и на сколько аминокислот?\n",
    "proteins = [\"P05067\", \"P10636\"]\n",
    "\n",
    "sequences = {}\n",
    "\n",
    "for p in proteins:\n",
    "    url = f\"https://www.ebi.ac.uk/proteins/api/proteins/{p}\"\n",
    "    response = requests.get(url, headers={\"Accept\": \"application/json\"})\n",
    "    data = response.json()\n",
    "    sequence = data[\"sequence\"][\"sequence\"]\n",
    "    sequences[p] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "711210c8-dbad-482e-aa58-aa85013d3002",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'Длина P05067:{len(sequences[\"P05067\"])}, длина P10636:{len(sequences[\"P10636\"])}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "088735a4-b7c1-4ac1-8e38-7b9f4475e313",
   "metadata": {},
   "source": [
    "Вывод: P05067 длиннее на 12 аминокислот"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e206cc8-1d7e-4db3-b8d3-57a5c114697c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#3 Получите данные о 5 различных белках (на ваш выбор) через API, используя их accession IDs. \n",
    "#Сохраните их entryName и длину последовательности в таблицу (data frame в R/Pandas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34ed19f-0a67-4e01-9239-21d26f20881c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d90c16ae-6267-462f-9129-4fa7f2fea692",
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions = [\"P68871\", \"P00734\", \"P01308\", \"P68133\", \"Q9Y6K9\"]\n",
    "result = [] #наш будущий датафрейм\n",
    "\n",
    "for accession in accessions:\n",
    "    response = requests.get(f\"https://www.ebi.ac.uk/proteins/api/proteins/{accession}\")\n",
    "    data = response.json()\n",
    "    entry_name = data.get('id', 'N/A')\n",
    "    sequence_length = len(data['sequence']['sequence'])\n",
    "    \n",
    "    result.append({\n",
    "        'accession': accession,\n",
    "        'entryName': entry_name,\n",
    "        'sequenceLength': sequence_length\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7fc9893-d7b0-4bf2-92ee-851a8e2bfd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(result)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f0f266-5234-4982-b85e-126ec6f9c8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4\n",
    "#(Python/R): Используя API PubMed, найдите 10 самых последних (сортировка по дате) \n",
    "#статей по запросу \"COVID-19 variants\". Выведите их ID и заголовки."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "691da77c-16a5-43a1-94bc-2b708e737d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "search_response = requests.get( #ищем статьи\n",
    "    \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\",\n",
    "    params={\n",
    "        'db': 'pubmed',\n",
    "        'term': '\"COVID-19 variants\"',\n",
    "        'retmode': 'json',\n",
    "        'retmax': 10,\n",
    "        'sort': 'pub+date'\n",
    "    }\n",
    ")\n",
    "search_data = search_response.json()\n",
    "ids = search_data['esearchresult']['idlist']\n",
    "\n",
    "fetch_response = requests.get( #получаем заголовки\n",
    "    \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\",\n",
    "    params={\n",
    "        'db': 'pubmed',\n",
    "        'id': ','.join(ids),\n",
    "        'retmode': 'json'\n",
    "    }\n",
    ")\n",
    "articles_data = fetch_response.json()\n",
    "\n",
    "for pubmed_id in ids:\n",
    "    article = articles_data['result'][pubmed_id]\n",
    "    print(f\"{pubmed_id}: {article.get('title', 'No title')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df3a7a3-0e9c-499e-921f-17754a9585bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5 (Python): Напишите функцию get_protein_name(accession), которая принимает на вход accession ID белка и возвращает его название (entryName). Обработайте случай, если белок не найден.\n",
    "def get_protein_name(accession):\n",
    "    response = requests.get(f\"https://www.ebi.ac.uk/proteins/api/proteins/{accession}\")\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        return data.get('protein', {}).get('recommendedName', {}).get('fullName', {}).get('value', 'Name not found')\n",
    "    else:\n",
    "        return f\"Error: Protein not found\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa64709-d596-4253-b7d9-aff52e6c12e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(get_protein_name(\"P04639\")) #пример работы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed312421-ccfd-423b-ae10-bdf352424931",
   "metadata": {},
   "outputs": [],
   "source": [
    "#6 (Python): Прочитайте из файла accessions.txt (нужно создать) список из 3-5 accession IDs.\n",
    "#Для каждого ID получите из Proteins API и запишите в новый файл его название и организм-источник."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c5de7f0-4c0e-4738-a5aa-b91c56cb16e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"accessions.txt\", \"w\") as f: #пзаполяем файлик\n",
    "    f.write(\"P04637\\nP05067\\nP10636\\nP68871\\nP00734\\n\")\n",
    "\n",
    "with open('accessions.txt', 'r') as file: #чтение айди\n",
    "    accessions = [line.strip() for line in file if line.strip()]\n",
    "\n",
    "with open('protein_info.txt', 'w') as output_file: #запись в файл\n",
    "    output_file.write(\"Accession\\tEntryName\\tOrganism\\n\")\n",
    "    \n",
    "    for accession in accessions:\n",
    "        response = requests.get(f\"https://www.ebi.ac.uk/proteins/api/proteins/{accession}\")\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            entry_name = data.get('id', 'N/A')\n",
    "            organism = data.get('organism', {}).get('names', [{}])[0].get('value', 'N/A')\n",
    "            output_file.write(f\"{accession}\\t{entry_name}\\t{organism}\\n\")\n",
    "\n",
    "print(\"Информация сохранена в protein_info.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62e7757c-db16-4550-b84b-aeaee958027c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7\n",
    "#(Python/R): Для найденных в задании 4 статей получите более подробную информацию, \n",
    "#используя Eutils API и ID статей (конечная точка efetch.fcgi). Извлеките имена авторов для каждой статьи.\n",
    "import xml.etree.ElementTree as ET\n",
    "search_response = requests.get(\n",
    "    \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\",\n",
    "    params={\n",
    "        'db': 'pubmed',\n",
    "        'term': '\"COVID-19 variants\"',\n",
    "        'retmode': 'json',\n",
    "        'retmax': 4\n",
    "    }\n",
    ")\n",
    "search_data = search_response.json()\n",
    "article_ids = search_data['esearchresult']['idlist']\n",
    "fetch_response = requests.get(\n",
    "    \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\",\n",
    "    params={\n",
    "        'db': 'pubmed',\n",
    "        'id': ','.join(article_ids),\n",
    "        'retmode': 'xml'\n",
    "    }\n",
    ")\n",
    "\n",
    "root = ET.fromstring(fetch_response.content)\n",
    "for article in root.findall('.//PubmedArticle'):\n",
    "    pmid = article.find('.//PMID').text\n",
    "    authors = article.findall('.//Author')\n",
    "    author_names = []\n",
    "    for author in authors:\n",
    "        last_name = author.find('LastName')\n",
    "        fore_name = author.find('ForeName')\n",
    "        if last_name is not None and fore_name is not None:\n",
    "            author_names.append(f\"{fore_name.text} {last_name.text}\")\n",
    "    \n",
    "    print(f\"ID: {pmid}\")\n",
    "    print(f\"Авторы: {', '.join(author_names)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7f23e69-cde2-4727-9391-738862d5b897",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 8 (Bash): Напишите команду curl, которая получает данные о белке, выполняет поиск по слову \"phosphorylation\" \n",
    "#(либо другое на ваш выбор) в ответе и выводит строки, содержащие это слово.\n",
    "!curl -s \"https://www.ebi.ac.uk/proteins/api/proteins/P04637\" | grep -i \"phosphorylation\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9697572c-ebe9-47e6-a71b-e14ba830a13f",
   "metadata": {},
   "outputs": [],
   "source": [
    "###9. Белок Q8WZ42 (Titin):  получить последовательность и разбить на строки по 80 символов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c1c03d-ce62-4af6-b18b-3dc8393abe74",
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -s \"https://www.ebi.ac.uk/proteins/api/proteins/Q8WZ42\" | grep -o '\"sequence\":\"[^\"]*\"' | cut -d'\"' -f4 | fold -w 80"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85c5f8e8-78e8-4070-801b-a7158127fc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 10\n",
    "library(httr)\n",
    "library(jsonlite)\n",
    "\n",
    "accessions <- c(\"P04637\", \"P05067\", \"P10636\", \"P03372\", \"Q8WZ42\", \"P01375\", \"P61586\", \"P62158\", \"P0DTD1\", \"P68871\")\n",
    "lengths <- c()\n",
    "names <- c()\n",
    "\n",
    "for (accession in accessions) {\n",
    "  response <- GET(paste0(\"https://www.ebi.ac.uk/proteins/api/proteins/\", accession))\n",
    "  data <- fromJSON(content(response, \"text\"))\n",
    "  lengths <- c(lengths, nchar(data$sequence$sequence))\n",
    "  names <- c(names, data$id)\n",
    "}\n",
    "\n",
    "barplot(lengths, names.arg = names, las = 2, main = \"Длина белков\", ylab = \"Аминокислоты\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea0484ec-9acf-4619-938f-42925d6d4134",
   "metadata": {},
   "outputs": [],
   "source": [
    "##11 не получилось"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56742604-d323-4b26-84d1-5400e0effa18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12. (Python): Ваш коллега написал код, который в цикле делает 100 запросов к API (по 1 белку). \n",
    "#Перепишите код, чтобы сделать всего 1 запрос, получающий данные сразу по 100 белкам (используйте соответствующий параметр в Proteins API)\n",
    "\n",
    "accessions = [\"P04637\", \"P05067\", \"P10636\", \"P03372\", \"Q8WZ42\"] #пример списка\n",
    "\n",
    "acc_str = \",\".join(accessions)\n",
    "url = \"https://www.ebi.ac.uk/proteins/api/proteins\"\n",
    "params = {\"accession\": acc_str}\n",
    "r = requests.get(url, params=params, headers={\"Accept\":\"application/json\"})\n",
    "data = r.json()\n",
    "data # data теперь содержит список результатов для всех accession\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "902f8c20-8191-4dab-9ecb-dfdd590ed487",
   "metadata": {},
   "outputs": [],
   "source": [
    "#13\n",
    "library(httr)\n",
    "acc <- \"P05067\"\n",
    "res <- GET(paste0(\"https://www.ebi.ac.uk/proteins/api/proteins/\", acc), add_headers(Accept=\"text/x-fasta\"))\n",
    "if (status_code(res)==200) {\n",
    "  writeBin(content(res, \"raw\"), paste0(acc, \".fasta\"))\n",
    "  cat(\"Saved\", paste0(acc, \".fasta\"), \"\\n\")\n",
    "} else {\n",
    "  cat(\"Error\", status_code(res), \"\\n\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3523d587-d695-485c-a2a8-f08990f52be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#14\n",
    "response = requests.get(\"https://www.ebi.ac.uk/proteins/api/nonexistent_endpoint\") #endpoint,который не существует\n",
    "\n",
    "if response.status_code == 404:\n",
    "    print(\"Ошибка 404: Белок не найден\")\n",
    "elif response.status_code == 200:\n",
    "    print(\"Запрос успешен\")\n",
    "else:\n",
    "    print(f\"Получен код ответа: {response.status_code}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3938c195-922f-40e1-8007-64a90569c181",
   "metadata": {},
   "outputs": [],
   "source": [
    "#16\n",
    "!curl -s \"https://www.ebi.ac.uk/proteins/api/proteins/P04637\" > protein_data.json #делаем файл"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ecf01b25-0bf3-4f71-9d9c-d8f0b9329664",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P04637\n"
     ]
    }
   ],
   "source": [
    "!jq -r '.accession' protein_data.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "05e37e0d-09cc-4857-9630-914e55040aa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Название: Pro-opiomelanocortin\n",
      "Ген: pomc\n",
      "Последние статьи: ['41196440', '41196411', '41195969']\n"
     ]
    }
   ],
   "source": [
    "#17\n",
    "#информация\n",
    "data1 = requests.get(\"https://www.ebi.ac.uk/proteins/api/proteins/P10000\").json()\n",
    "print(\"Название:\", data1['protein']['recommendedName']['fullName']['value'])\n",
    "print(\"Ген:\", data1['gene'][0]['name']['value'])\n",
    "\n",
    "#поиск статей\n",
    "pubmed_data = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=APP[Gene]&retmode=json&retmax=5\").json()\n",
    "article_ids = pubmed_data['esearchresult']['idlist']\n",
    "print(\"Последние статьи:\", article_ids[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7e220eb6-f854-4315-b2aa-8ce265805261",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/eyzorina/Учеба/homeworks\n",
      "=== Текущий статус ===\n",
      "On branch dev\n",
      "Your branch is ahead of 'origin/dev' by 2 commits.\n",
      "  (use \"git push\" to publish your local commits)\n",
      "\n",
      "nothing to commit, working tree clean\n",
      "\n",
      "=== Обновляем main ===\n",
      "Switched to branch 'main'\n",
      "Your branch is up to date with 'origin/main'.\n",
      "remote: Enumerating objects: 1, done.\u001b[K\n",
      "remote: Counting objects: 100% (1/1), done.\u001b[K\n",
      "remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)\u001b[K\n",
      "Unpacking objects: 100% (1/1), 890 bytes | 445.00 KiB/s, done.\n",
      "From https://github.com/evdokiazorina/homeworks\n",
      " * branch            main       -> FETCH_HEAD\n",
      "   d9a7bbc..ee40221  main       -> origin/main\n",
      "Updating d9a7bbc..ee40221\n",
      "Fast-forward\n",
      " Homework4.R |  45 \u001b[32m+++++++++++++++++++++++++\u001b[m\n",
      " Homework5.R | 107 \u001b[32m++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\u001b[m\n",
      " homework3   |  67 \u001b[32m++++++++++++++++++++++\u001b[m\u001b[31m---------------\u001b[m\n",
      " 3 files changed, 192 insertions(+), 27 deletions(-)\n",
      " create mode 100644 Homework4.R\n",
      " create mode 100644 Homework5.R\n"
     ]
    }
   ],
   "source": [
    "%cd homeworks\n",
    "\n",
    "\n",
    "print(\"=== Текущий статус ===\")\n",
    "!git status\n",
    "\n",
    "\n",
    "!git checkout main\n",
    "!git pull origin main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b2d8cd82-cf3d-4158-800b-e2b62666067f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fatal: not a git repository (or any of the parent directories): .git\n",
      "fatal: not a git repository (or any of the parent directories): .git\n"
     ]
    }
   ],
   "source": [
    "!git checkout main\n",
    "!git pull origin main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf4dfd8f-62ad-49bd-a43c-de25e2dc62b9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
